### PR TITLE
[Sprint: 47] XD-2814: Durable Pub/Sub with Rabbit Bus

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/XD-2814

The `RabbitMessageBus` now supports durable queues for pub/sub named channels
(`tap:` and `topic:`). Such queues are eligible for automatic DLQ creation
and binding.

This change also means that multiple instances of the first module in a stream
when the input channel is a named pub/sub are now competing consumers for
messages in that channel.

Previously, each instance received all messages which was incorrect (XD-2115).

XD-2814: Initial Commit

Use Capability

Tests and DLQ Setup

Docs

__this PR will conflict with #1547 and will need to be rebased/resolved if that is merged first (or vice versa)__

### CODE REVIEW

**Comment by garyrussell** [comment]
Rebased; resolved doc conflicts after merge of #1547

File: extensions/spring-xd-extension-http/Californium.properties [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#Californium CoAP Properties file
```
Comment (ilayaperumalg) [inline extensions/spring-xd-extension-http/Californium.properties]: I think this CoAP properties file intended to be checked in with a different PR ;)

File: spring-xd-dirt/Californium.properties [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#Californium CoAP Properties file
```
Comment (ilayaperumalg) [inline spring-xd-dirt/Californium.properties]: Same here.

File: src/docs/asciidoc/Deployment.asciidoc [hunk 3]
```diff
@@ -94,6 +94,8 @@ module.[modulename].consumer.replyHeaderPatterns:: Controls which message header
 module.[modulename].consumer.requeue:: Whether messages will be requeued (and retried) on failure *(default true)*
 module.[modulename].consumer.transacted:: Whether consumers use transacted channels *(default false)*
 module.[modulename].consumer.txSize:: The number of delivered messages between acknowledgements (when _ackMode=AUTO_) *(default 1)*
+module.[modulename].consumer.durable:: When true, publish/subscribe named channels (`tap:`, `topic:`) will be backed by a durable queue and will be eligible for dead-letter configuration, accoring to the `autBindDLQ` setting.
```
Comment (ilayaperumalg) [inline src/docs/asciidoc/Deployment.asciidoc]: Typo on the name of the property. `durable` -> `durableSubscription`.

File: src/docs/asciidoc/MessageBus.asciidoc [hunk 4]
```diff
@@ -56,13 +56,23 @@ There is no automated mechanism provided to move dead lettered messages back to
...
+This property is similar to a JMS durable subscription to a topic and is `false` by default.
+When `false` (default), the queue backing such a named channel is declared `auto-delete` and is removed when the stream is undeployed.
+A DLQ will not be created for such queues.
+When `true`, the queue becomes permanent (durable) and is not removed when the stream is undeployed.
+Also, when `true`, the queue is eligible for DLQ provisioning, according to the `autoBindDLQ` deployment property.
+`durable` can be set at the bus level, or in an individual deployment property, such as:
+
+`stream create ticktock --definition="time --fixedDelay=5 | log" --deploy`
+`stream create tttap --definition="tap:stream:ticktock > log"`
+`stream deploy tttap --properties=module.log.consumer.durable=true`
```
Comment (ilayaperumalg) [inline src/docs/asciidoc/MessageBus.asciidoc]: `durable` -> `durableSubscription`

**Comment by ilayaperumalg** [comment]
@garyrussell I see some issues when deploying the stream with module count `0`.

I have one admin and three containers running:
My stream definitions are:
----
stream create a1 --definition "http | log" --deploy
stream create a2 --definition "tap:stream:a1 > log"
stream deploy a2 --properties "module.*.count=0"

----

Following exception is thrown at the admin:
----
INFO DeploymentSupervisor-0 zk.ZKStreamDeploymentHandler - Deployment status for stream 'a2': DeploymentStatus{state=deployed,error(s)=org.springframework.context.ApplicationContextException: Failed to start bean 'org.springframework.integration.monitor.IntegrationMBeanExporter#0#d43dd0f4-0dab-4a66-87ff-8bd394218443'; nested exception is org.springframework.jmx.export.UnableToRegisterMBeanException: Unable to register MBean [MessageChannelMonitor: [name=input, sends=0]] with key 'org.springframework.integration:type=MessageChannel,name=input'; nested exception is javax.management.InstanceAlreadyExistsException: xd.a2:module=log.0,component=MessageChannel,name=input
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:176)
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:51)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:346)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:149)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:112)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:770)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:483)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:691)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:139)
	at org.springframework.xd.module.core.SimpleModule.initialize(SimpleModule.java:214)
	at org.springframework.xd.dirt.module.ModuleDeployer.doDeploy(ModuleDeployer.java:217)
	at org.springframework.xd.dirt.module.ModuleDeployer.deploy(ModuleDeployer.java:200)
	at org.springframework.xd.dirt.server.container.DeploymentListener.deployModule(DeploymentListener.java:365)
	at org.springframework.xd.dirt.server.container.DeploymentListener.deployStreamModule(DeploymentListener.java:334)
	at org.springframework.xd.dirt.server.container.DeploymentListener.onChildAdded(DeploymentListener.java:183)
	at org.springframework.xd.dirt.server.container.DeploymentListener.childEvent(DeploymentListener.java:149)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache$5.apply(PathChildrenCache.java:509)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache$5.apply(PathChildrenCache.java:503)
	at org.apache.curator.framework.listen.ListenerContainer$1.run(ListenerContainer.java:92)
	at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
	at org.apache.curator.framework.listen.ListenerContainer.forEach(ListenerContainer.java:83)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache.callListeners(PathChildrenCache.java:500)
	at org.apache.curator.framework.recipes.cache.EventOperation.invoke(EventOperation.java:35)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache$10.run(PathChildrenCache.java:762)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.springframework.jmx.export.UnableToRegisterMBeanException: Unable to register MBean [MessageChannelMonitor: [name=input, sends=0]] with key 'org.springframework.integration:type=MessageChannel,name=input'; nested exception is javax.management.InstanceAlreadyExistsException: xd.a2:module=log.0,component=MessageChannel,name=input
	at org.springframework.jmx.export.MBeanExporter.registerBeanNameOrInstance(MBeanExporter.java:614)
	at org.springframework.integration.monitor.IntegrationMBeanExporter.registerChannels(IntegrationMBeanExporter.java:735)
	at org.springframework.integration.monitor.IntegrationMBeanExporter.doStart(IntegrationMBeanExporter.java:452)
	at org.springframework.integration.monitor.IntegrationMBeanExporter.start(IntegrationMBeanExporter.java:403)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:173)
	... 31 more
Caused by: javax.management.InstanceAlreadyExistsException: xd.a2:module=log.0,component=MessageChannel,name=input
	at com.sun.jmx.mbeanserver.Repository.addMBean(Repository.java:437)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerWithRepository(DefaultMBeanServerInterceptor.java:1898)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerDynamicMBean(DefaultMBeanServerInterceptor.java:966)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerObject(DefaultMBeanServerInterceptor.java:900)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.registerMBean(DefaultMBeanServerInterceptor.java:324)
	at com.sun.jmx.mbeanserver.JmxMBeanServer.registerMBean(JmxMBeanServer.java:522)
	at org.springframework.jmx.support.MBeanRegistrationSupport.doRegister(MBeanRegistrationSupport.java:195)
	at org.springframework.jmx.export.MBeanExporter.registerBeanInstance(MBeanExporter.java:667)
	at org.springframework.jmx.export.MBeanExporter.registerBeanNameOrInstance(MBeanExporter.java:610)
	... 35 more
; org.springframework.amqp.AmqpIOException: java.io.IOException
	at org.springframework.amqp.rabbit.support.RabbitExceptionTranslator.convertRabbitAccessException(RabbitExceptionTranslator.java:63)
	at org.springframework.amqp.rabbit.connection.RabbitAccessor.convertRabbitAccessException(RabbitAccessor.java:110)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:1051)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1028)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1004)
	at org.springframework.amqp.rabbit.core.RabbitAdmin.declareQueue(RabbitAdmin.java:148)
	at org.springframework.xd.dirt.integration.rabbit.RabbitMessageBus.declareQueueIfNotPresent(RabbitMessageBus.java:684)
	at org.springframework.xd.dirt.integration.rabbit.RabbitMessageBus.bindPubSubConsumer(RabbitMessageBus.java:421)
	at org.springframework.xd.dirt.plugins.AbstractMessageBusBinderPlugin.bindMessageConsumer(AbstractMessageBusBinderPlugin.java:274)
	at org.springframework.xd.dirt.plugins.AbstractMessageBusBinderPlugin.bindConsumerAndProducers(AbstractMessageBusBinderPlugin.java:155)
	at org.springframework.xd.dirt.plugins.stream.StreamPlugin.postProcessModule(StreamPlugin.java:73)
	at org.springframework.xd.dirt.module.ModuleDeployer.postProcessModule(ModuleDeployer.java:238)
	at org.springframework.xd.dirt.module.ModuleDeployer.doDeploy(ModuleDeployer.java:218)
	at org.springframework.xd.dirt.module.ModuleDeployer.deploy(ModuleDeployer.java:200)
	at org.springframework.xd.dirt.server.container.DeploymentListener.deployModule(DeploymentListener.java:365)
	at org.springframework.xd.dirt.server.container.DeploymentListener.deployStreamModule(DeploymentListener.java:334)
	at org.springframework.xd.dirt.server.container.DeploymentListener.onChildAdded(DeploymentListener.java:183)
	at org.springframework.xd.dirt.server.container.DeploymentListener.childEvent(DeploymentListener.java:149)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache$5.apply(PathChildrenCache.java:509)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache$5.apply(PathChildrenCache.java:503)
	at org.apache.curator.framework.listen.ListenerContainer$1.run(ListenerContainer.java:92)
	at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
	at org.apache.curator.framework.listen.ListenerContainer.forEach(ListenerContainer.java:83)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache.callListeners(PathChildrenCache.java:500)
	at org.apache.curator.framework.recipes.cache.EventOperation.invoke(EventOperation.java:35)
	at org.apache.curator.framework.recipes.cache.PathChildrenCache$10.run(PathChildrenCache.java:762)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException
	at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:106)
	at com.rabbitmq.client.impl.AMQChannel.wrap(AMQChannel.java:102)
	at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:124)
	at com.rabbitmq.client.impl.ChannelN.queueDeclare(ChannelN.java:833)
	at com.rabbitmq.client.impl.ChannelN.queueDeclare(ChannelN.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$CachedChannelInvocationHandler.invoke(CachingConnectionFactory.java:625)
	at com.sun.proxy.$Proxy95.queueDeclare(Unknown Source)
	at org.springframework.amqp.rabbit.core.RabbitAdmin.declareQueues(RabbitAdmin.java:448)
	at org.springframework.amqp.rabbit.core.RabbitAdmin.access$200(RabbitAdmin.java:54)
	at org.springframework.amqp.rabbit.core.RabbitAdmin$3.doInRabbit(RabbitAdmin.java:151)
	at org.springframework.amqp.rabbit.core.RabbitAdmin$3.doInRabbit(RabbitAdmin.java:148)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:1045)
	... 30 more
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=405, reply-text=RESOURCE_LOCKED - cannot obtain exclusive access to locked queue 'xdbus.a2.tap:stream:a1.http.0' in vhost '/', class-id=50, method-id=10)
	at com.rabbitmq.utility.ValueOrException.getValue(ValueOrException.java:67)
	at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:33)
	at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:343)
	at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:216)
	at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:118)
	... 43 more
Caused by: com.rabbitmq.client.ShutdownSignalException: channel error; protocol method: #method<channel.close>(reply-code=405, reply-text=RESOURCE_LOCKED - cannot obtain exclusive access to locked queue 'xdbus.a2.tap:stream:a1.http.0' in vhost '/', class-id=50, method-id=10)
	at com.rabbitmq.client.impl.ChannelN.asyncShutdown(ChannelN.java:478)
	at com.rabbitmq.client.impl.ChannelN.processAsync(ChannelN.java:315)
	at com.rabbitmq.client.impl.AMQChannel.handleCompleteInboundCommand(AMQChannel.java:144)
	at com.rabbitmq.client.impl.AMQChannel.handleFrame(AMQChannel.java:91)
	at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:552)
	... 1 more

----

**Comment by garyrussell** [comment]
@ilayaperumalg pushed - fixed the exlusive queue problem; I also removed the bogus CoAP files.

Note that `BusProperties.DURABLE` is, in fact, `durable`, not `durableSubscription` - I thought a shorter property would be better for the command line.

However, the bus-wide setting (in application.yml) is `durableSubscription`.

If you prefer, I can make these consistent.

**Comment by ilayaperumalg** [comment]
@garyrussell thanks for fixing the issue. I see why you didn't name it as `durable` in the centralized config. But I think having two different property names might confuse the user (though we clear that in the documentation). I would prefer having `durableSubscription` in both places to make it consistent.

**Comment by ilayaperumalg** [comment]
One observation with this PR change is that, when the pub/sub consumer queue is made durable by the consumer module (`module.moduleName.consumer.durableSubscription` set to true) and this module gets undeployed and re-deployed without durable option (`module.moduleName.consumer.durableSubscription`), the consumer queue would still remain `durable`. Since the queue is not destroyed upon undeploy, this is expected. To workaround, one has to use rabbit bus cleaner util to clean the queue and re-deploy.

**Comment by garyrussell** [comment]
Good observation - yes, RabbitMQ doesn't allow a queue's characteristics to be changed - the queue has to be deleted and re-created.

Perhaps we should make a note in the docs?

**Comment by ilayaperumalg** [comment]
Sure, I can update the doc while merging:

how about this:

---
module.[modulename].consumer.durableSubscription:: When true, publish/subscribe named channels (`tap:`, `topic:`) will be backed by a durable queue and will be eligible for dead-letter configuration, accoring to the `autBindDLQ` setting.
Also, when this is set to true and the module needs to be re-deployed manually with `false` option (non-durable queue), the queue needs to be deleted and let the re-deployment create a non-durable queue. Similarly, when this property is set from `false` to `true` upon re-deployment, the non-durable queue needs to be deleted before re-deployment.

----

**Comment by garyrussell** [comment]
Looks good, but going from non-durable to durable is not a problem...

Perhaps instead of...

>Also, when this is set to true and the module needs to be re-deployed manually with `false` option (non-durable queue), the queue needs to be deleted and let the re-deployment create a non-durable queue. Similarly, when this property is set from false to true upon re-deployment, the non-durable queue needs to be deleted before re-deployment.

Note that, since RabbitMQ doesn't permit queue attributes to be changed, changing the `durableSubscription` property from `true` to `false` between deployments, without first removing the queue, will not have any effect. If a stream is deployed with `durableSubscription=true`, and you wish to change it to a non-durable subscription, you will need to remove the queue from RabbitMQ before redeploying. Spring XD will create the queue the with the appropriate settings, unless the queue exists already. Changing from a non-durable subscription to a durable subscription will not have this problem because, for a non-durable subscription, the queue will be automatically deleted when the stream is undeployed.

**Comment by ilayaperumalg** [comment]
yeah, I missed that part about going from non-durable to durable :-). will update the doc.

**Comment by ilayaperumalg** [comment]
LGTM, merging.

**Comment by ilayaperumalg** [comment]
Changed the property name `durableSubscriptions` to be consistent with BusProperties and Centralized properties. Updated the doc as diccussed above.

Squashed, rebased and merged as 8992ec09c7effb968f73033ff3ee3b6ba977b1de.
